### PR TITLE
switch to boto3 and add support for application ELBs with target groups.

### PR DESCRIPTION
**Pull request by jonhadfield**
##### ISSUE TYPE
 - Feature Pull Request

##### COMPONENT NAME
ec2_asg

##### ANSIBLE VERSION
```
ansible 2.2.0.0 (detached HEAD cdec853e37) last updated 2016/11/01 12:32:01 (GMT +100)
  lib/ansible/modules/core: (detached HEAD fe9c56a003) last updated 2016/11/01 12:32:05 (GMT +100)
  lib/ansible/modules/extras: (detached HEAD f564e77a08) last updated 2016/11/01 12:32:05 (GMT +100)
  config file =
  configured module search path = Default w/o overrides
```

##### SUMMARY
- Upgrade module to use boto3 in order to support new functionality.
- Added support for Application ELBs via Target Groups

Compliments the following PRs in that it enables creation of an ASG with ansible generated application load balancers and target groups:
elb_target_group: https://github.com/ansible/ansible-modules-extras/pull/2947
elb_application_lb: https://github.com/ansible/ansible-modules-extras/pull/2946

Note: 
- Only significant logic change is to support the integration with target groups/application lbs, so any existing issues remain
- Only minimal testing performed

### CODE REVIEW

**Comment by gregdek** [comment]
Thanks @jonhadfield for this PR. This PR requires revisions, either because it fails to build or by reviewer request. Please make the suggested revisions. When you are done, please comment with text 'ready_for_review' and we will put this PR back into review.

[This message brought to you by your friendly Ansibull-bot.]

File: cloud/amazon/ec2_asg.py
Comment (ryansb) [inline cloud/amazon/ec2_asg.py]: This module was added in 1.6, don't change this.

File: cloud/amazon/ec2_asg.py
Comment (ryansb) [inline cloud/amazon/ec2_asg.py:56]: This functionality was just added in 2.3, please pull this back in.

File: cloud/amazon/ec2_asg.py
Comment (ryansb) [inline cloud/amazon/ec2_asg.py:151]: Same here, this parameter was added in 2.3

File: cloud/amazon/ec2_asg.py
Comment (ryansb) [inline cloud/amazon/ec2_asg.py:231]: The module's failure modes depend on this to send back the full trace to the user.

File: cloud/amazon/ec2_asg.py
Comment (ryansb) [inline cloud/amazon/ec2_asg.py]: I think boto3 changes the logger name here, so I think this line is obsolete.

File: cloud/amazon/ec2_asg.py
Comment (ryansb) [inline cloud/amazon/ec2_asg.py]: This change seems to be style-only, can you revert it since you aren't changing any actual code on this line?

File: cloud/amazon/ec2_asg.py
Comment (ryansb) [inline cloud/amazon/ec2_asg.py:404]: Same as above: please don't remove this functionality

File: cloud/amazon/ec2_asg.py
Comment (ryansb) [inline cloud/amazon/ec2_asg.py:567]: Instead of just the try/except, can you use the new AWSRetry library that's in ec2's module_utils for external calls like this? It handles backoffs/retries automatically.

**Review by ryansb — CHANGES_REQUESTED** [review]
Thanks for this PR, but it needs a few changes (and a review from @garethr, the original author) before it can be merged in.

**Comment by jonhadfield** [comment]
I had major issues trying to get this PR moved, so I've created a replacement PR at: https://github.com/ansible/ansible/pull/19667
@ryansb I've implemented the missing placement group and suspended processes functionality using boto3. It was missing due to my fork being older than when those changes were merged.
I've un-PEP8'd a lot of the code so it's more simple to compare and doesn't confuse attribution.
The only bit I've not yet incorporated is the use of AWSRetry as I don't have time this week.

**Pull request closed without merging**
